src/crimson/os/seastore/extent_placement_manager.h

```diff
@@ -186,43 +198,61 @@ class ExtentPlacementManager {
 
     // XXX: bp might be extended to point to differnt memory (e.g. PMem)
     // according to the allocator.
-    auto bp = ceph::bufferptr(
+    auto alloc_paddr = [this](reclaim_gen_t gen, 
+      data_category_t category, extent_len_t length) 
+      -> alloc_result_t {
+      auto bp = ceph::bufferptr(
       buffer::create_page_aligned(length));
-    bp.zero();
+      bp.zero();
+      paddr_t addr;
+      if (category == data_category_t::DATA) {
+	if (data_writers_by_gen.size() == 0 || !data_writers_by_gen[gen]) {
+	  addr = make_record_relative_paddr(0);
+	} else {
+	  addr = data_writers_by_gen[gen]->alloc_paddr(length);
+	}
+      } else {
+	if (md_writers_by_gen.size() == 0 || gen == 0 || !md_writers_by_gen[gen]) {
+	  addr = make_record_relative_paddr(0);
+	} else {
+	  addr = md_writers_by_gen[gen - 1]->alloc_paddr(length);
+	}
+      }
+      return {addr,
+	      std::move(bp),
+	      gen};
+    };
 
     if (!is_logical_type(type)) {
       // TODO: implement out-of-line strategy for physical extent.
-      return {make_record_relative_paddr(0),
-              std::move(bp),
-              0};
+      return alloc_paddr(0, data_category_t::METADATA, length);
```

Comment:
We can safely assert that all physical extents are METADATA before passing the category into the alloc method.